### PR TITLE
Parser can restart on named formula errors

### PR DESCRIPTION
**Pull request by rick-nguyen**
Changes identifier lexer with semicolon as new delimiter end
Changes parser to add an error name formula to the list for a given identifier once, and continue parsing as normal after next semicolon (or terminates if we hit Eof)

### CODE REVIEW

File: src/libraries/Microsoft.PowerFx.Core/Lexer/TexlLexer.cs [hunk 1]
```diff
@@ -517,6 +517,9 @@ public static bool IsSimpleIdentCh(char ch)
         // Returns true if the specified character is a new line character.
         public static bool IsNewLineCharacter(char ch) => ch == '\n';
 
+        // Returns true if the specified character is a semicolon character.
+        public static bool IsSemicolonCharacter(char ch) => ch == ';';
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Lexer/TexlLexer.cs:521]: Don't compare to semicolon, generally. It's `;;` in some locales as well. 

 TexlLexer.TryGetPunctuator(str, out tid) && tid = TokKind.Semicolon

File: src/libraries/Microsoft.PowerFx.Core/Lexer/TexlLexer.cs [hunk 2]
```diff
@@ -1275,6 +1278,13 @@ private string LexIdentCore(out bool fDelimiterStart, out bool fDelimiterEnd)
                         fDelimiterEnd = false;
                         break;
                     }
+                    else if (IsSemicolonCharacter(CurrentChar))
+                    {
+                        // Terminate an identifier on a semicolon character
```
[the remark this answers is not in the document]
Comment (lesaltzm) [reply src/libraries/Microsoft.PowerFx.Core/Lexer/TexlLexer.cs:1283]: Likely that involves storing some kind of start index for paired things like ident tokens, and restarting at that if we hit EOF with errors

File: src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs [hunk 3]
```diff
@@ -642,7 +651,12 @@ private TexlNode ParseExpr(Precedence precMin, TexlNode node = null)
                         case TokKind.StrLit:
                         case TokKind.True:
                         case TokKind.False:
-                            PostError(_curs.TokCur, TexlStrings.ErrOperatorExpected);
+                            PostError(_curs.TokCur, TexlStrings.ErrOperatorExpected);
+                            if (_flagsMode.Peek().HasFlag(Flags.NamedFormulas))
+                            {
+                                goto case TokKind.Semicolon;
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs:657]: yikes. Why?

File: src/tests/Microsoft.PowerFx.Core.Tests/ParseTests.cs [hunk 4]
```diff
@@ -769,5 +769,17 @@ internal void TestFormulasParseError(string script)
 
             Assert.True(result.HasError);
         }
+
+        [Theory]
+        [InlineData("a = 10; b = in'valid ; c = 20;", "c")]
+        [InlineData("a = 10; b = in'valid", "a")]
```
Comment (lesaltzm) [inline src/tests/Microsoft.PowerFx.Core.Tests/ParseTests.cs:775]: also tests for other sources  of parser errors

File: src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs [hunk 5]
```diff
@@ -866,6 +886,9 @@ private TexlNode ParseOperand()
                     return new SelfNode(ref _idNext, _curs.TokMove());
 
                 case TokKind.Eof:
+                    return CreateError(_curs.TokCur, TexlStrings.ErrOperandExpected);
+
+                case TokKind.Semicolon:
```
Comment (rick-nguyen) [inline src/libraries/Microsoft.PowerFx.Core/Parser/TexlParser.cs:891]: I added another case to ParseOperand for semicolon. If we hit a semicolon token when we expect an operand to parse, we just return the error (without moving the token) and stop recursing, and continue parsing the rest of the named formula as normal.

File: src/libraries/Microsoft.PowerFx.Core/Lexer/TexlLexer.cs [hunk 6]
```diff
@@ -1275,6 +1288,13 @@ private string LexIdentCore(out bool fDelimiterStart, out bool fDelimiterEnd)
                         fDelimiterEnd = false;
                         break;
                     }
+                    else if (_lex.TryGetPunctuator(CurrentChar.ToString(), out var tid) && tid == TokKind.Semicolon)
+                    {
+                        // Don't know if semicolon is end of identifier so we will store for fallback
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Lexer/TexlLexer.cs:1293]: add a comment that this is about enabling parser restarts

File: src/libraries/Microsoft.PowerFx.Core/Lexer/TexlLexer.cs [hunk 7]
```diff
@@ -1275,6 +1288,13 @@ private string LexIdentCore(out bool fDelimiterStart, out bool fDelimiterEnd)
                         fDelimiterEnd = false;
                         break;
                     }
+                    else if (_lex.TryGetPunctuator(CurrentChar.ToString(), out var tid) && tid == TokKind.Semicolon)
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Lexer/TexlLexer.cs:1291]: this won't quite work, add a test in a `;;` separator locale and you'll see the issue.

File: src/tests/Microsoft.PowerFx.Core.Tests/ParseTests.cs [hunk 8]
```diff
@@ -769,5 +769,42 @@ internal void TestFormulasParseError(string script)
...
+        [InlineData("a = 10; b = 3213d 123123asdf; c = 23;", "c")]
+        [InlineData("a = 10; b = 3213d 123123asdf;; c = 23;", "c")]
+        [InlineData("a = 10; b = 321;3;d ;;;123123asdf;; c = 23;", "c")]
+        public void TestFormulaParseRestart(string script, string key)
+        {
+            var formulasResult = TexlParser.ParseFormulasScript(script);
+            Assert.True(formulasResult.HasError);
+
+            // Parser restarted, and found 'c' correctly
+            Assert.Contains(formulasResult.NamedFormulas, kvp => kvp.Key.Name.Value == key);
```
Comment (lesaltzm) [inline src/tests/Microsoft.PowerFx.Core.Tests/ParseTests.cs:796]: some things to consider: when we do this, what errors are in the set of errors we return?

What if we have multiple restarts in the same expression? i.e

"a = 10; b = in'valid ; c = 20; d = also(invalid; e = 44;"

Comment (lesaltzm) [reply src/tests/Microsoft.PowerFx.Core.Tests/ParseTests.cs:796]: could be handled in a future PR, just something to consider.

File: src/libraries/Microsoft.PowerFx.Core/Lexer/TexlLexer.cs [hunk 9]
```diff
@@ -1275,6 +1288,25 @@ private string LexIdentCore(out bool fDelimiterStart, out bool fDelimiterEnd)
...
+                    else if (CurrentChar.ToString() == PunctuatorSemicolonInvariant)
+                    {
+                        // This is to enable parser restarting
+                        // Don't know if semicolon is end of identifier so we will store for fallback
+                        // Some locales use ;;, so we check for match and move shift 2 characters instead of 1
+                        if (_lex.TryGetPunctuator(CurrentChar.ToString(), out TokKind tid) && tid == TokKind.Semicolon)
+                        {
+                            semicolonIndex = semicolonIndex == -1 ? CurrentPos : semicolonIndex;
+                        }
+                        else if (_lex.TryGetPunctuator(CurrentChar.ToString() + _text[Math.Min(CurrentPos + 1, _charCount)].ToString(), out tid) && tid == TokKind.Semicolon)
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Lexer/TexlLexer.cs:1300]: @bryancroteau-MSFT any thoughts on a better way to do this?

**Review by lesaltzm — DISMISSED** [review]

**Comment by LucGenetier** [comment]
✅ No public API change.

**Pull request merged**